src/base/UnixSocketHandler.cpp

```diff
@@ -159,7 +159,9 @@ void UnixSocketHandler::close(int fd) {
   auto m = it->second;
   lock_guard<std::recursive_mutex> guard(*m);
   VLOG(1) << "Closing connection: " << fd;
-  FATAL_FAIL(::close(fd));
+  if (errno != EBADF && errno != EPIPE) {
```

Comment:
errno is based on the most recently called system function, which is undefined in this case because close hasn't been called yet.  Can you provide another way to check if the fd is valid?

Reply:
Yes, you're right. In my case, close() was being called after a failed read(), but there are definitely code paths that don't call system functions before the close. I have updated the pull request to use fcntl() to check for a valid FD. I only check the return value of fcntl(), because in testing on iOS devices, I was able to get it to fail with at least 5 different errno values, depending on when and how the connection was terminated.